src/Handler/StreamHandler.php

```diff
@@ -301,21 +302,44 @@ private function createStream(RequestInterface $request, array $options)
             );
         }
 
+        $uri = $this->resolveUri($request->getUri(), $options);
+
         $context = $this->createResource(
             function () use ($context, $params) {
                 return stream_context_create($context, $params);
             }
         );
 
         return $this->createResource(
-            function () use ($request, &$http_response_header, $context) {
-                $resource = fopen((string) $request->getUri()->withFragment(''), 'r', null, $context);
+            function () use ($uri, &$http_response_header, $context) {
+                $resource = fopen((string) $uri, 'r', null, $context);
                 $this->lastHeaders = $http_response_header;
                 return $resource;
             }
         );
     }
 
+    private function resolveUri(UriInterface $uri, array $options)
```

Comment:
Maybe call it resolveIp or resolveAddress? Since that's what we do here, isn't it?